astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+            Radial velocity of target.
+
+        Notes
+        -----
+        This is different from the ``.radial_velocity`` property of a
+        coordinate frame in that this calculates the radial velocity with
+        respect to the *observer*, not the origin of the frame.
+        """
+        return self._calculate_radial_velocity(self._observer, self._target,
+                                               as_scalar=True)
+
+    @property
+    def redshift(self):
+        """
+        Redshift of target relative to observer. Calculated from the radial
+        velocity.
+
+        Returns
+        -------
+        float
+            Redshift of target.
+        """
+        return _velocity_to_redshift(self.radial_velocity)
+
+    @staticmethod
+    def _calculate_radial_velocity(observer, target, as_scalar=False):
+        """
+        Compute the line-of-sight velocity from the observer to the target.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame`
+            The frame of the observer.
+        target : `~astropy.coordinates.BaseCoordinateFrame`
+            The frame of the target.
+        as_scalar : bool
+            If `True`, the magnitude of the velocity vector will be returned,
+            otherwise the full vector will be returned.
+
+        Returns
+        -------
+        `~astropy.units.Quantity`
+            The radial velocity of the target with respect to the observer.
+        """
+
+        # Convert observer and target to ICRS to avoid finite differencing
+        # calculations that lack numerical precision.
+        observer_icrs = observer.transform_to(ICRS)
+        target_icrs = target.transform_to(ICRS)
+
+        pos_hat = SpectralCoord._norm_d_pos(observer_icrs, target_icrs)
+
+        d_vel = target_icrs.velocity - observer_icrs.velocity
+
+        vel_mag = np.dot(d_vel.d_xyz, pos_hat.xyz)
+
+        if NUMPY_LT_1_17:
+            vel_mag *= d_vel.d_xyz.unit
+
+        if as_scalar:
+            return vel_mag
+        else:
+            return vel_mag * pos_hat.xyz
+
+    @staticmethod
+    def _norm_d_pos(observer, target):
+        """
+        Calculate the normalized position vector between two frames.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The observation frame or coordinate.
+        target : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+            The target frame or coordinate.
+
+        Returns
+        -------
+        pos_hat : `BaseRepresentation`
+            Position representation.
+        """
+        d_pos = (target.data.without_differentials() -
+                 observer.data.without_differentials()).to_cartesian()
+
+        dp_norm = d_pos.norm()
+
+        # Reset any that are 0 to 1 to avoid nans from 0/0
+        dp_norm.ravel()[dp_norm.ravel() == 0] = 1 * dp_norm.unit
```

Comment:
No need to `ravel`, just `dp_norm[dp_norm == 0] = 1 * dp_norm.unit`

Reply:
Fixed